File: gitdot-web/packages/gitdot-client/src/review.ts
import "server-only";

import type {
  PublishReviewRequest,
  ReplyToReviewCommentRequest,
  ReviewReviewDiffRequest,
  UpdateReviewCommentRequest,
  UpdateReviewDiffRequest,
  UpdateReviewRequest,
} from "gitdot-api";
import {
  GetReviewDiffResponse,
  ListReviewsResponse,
  ReviewCommentResource,
  ReviewerResource,
  ReviewResource,
} from "gitdot-api";
import {
  authDelete,
  authFetch,
  authPatch,
  authPost,
  GITDOT_SERVER_URL,
  handleEmptyResponse,
  handleResponse,
  toQueryString,
} from "./util";

export async function listReviews(
  owner: string,
  repo: string,
  opts?: { cursor?: string; limit?: number },
): Promise<ListReviewsResponse | null> {
  const qs = toQueryString({ cursor: opts?.cursor, limit: opts?.limit });
  const url = `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/reviews${qs ? `?${qs}` : ""}`;
  const response = await authFetch(url);
  return await handleResponse(response, ListReviewsResponse);
}

export async function getReview(
  owner: string,
  repo: string,
  number: number | string,
): Promise<ReviewResource | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}`,
  );

  return await handleResponse(response, ReviewResource);
}

export async function getReviewDiff(
  owner: string,
  repo: string,
  number: number | string,
  position: number,
  revision?: number,
  compareTo?: number,
): Promise<GetReviewDiffResponse | null> {
  const params: Record<string, number> = {};
  if (revision !== undefined) params.revision = revision;
  if (compareTo !== undefined) params.compare_to = compareTo;
  const query = toQueryString(params);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}${query ? `?${query}` : ""}`,
  );

  return await handleResponse(response, GetReviewDiffResponse);
}

export async function addReviewer(
  owner: string,
  repo: string,
  number: number,
  request: { user_name: string },
): Promise<ReviewerResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/reviewer`,
    request,
  );

  return await handleResponse(response, ReviewerResource);
}

export async function removeReviewer(
  owner: string,
  repo: string,
  number: number,
  reviewerName: string,
): Promise<void> {
  const response = await authDelete(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/reviewer/${reviewerName}`,
  );

  await handleEmptyResponse(response);
}

export async function updateDiff(
  owner: string,
  repo: string,
  number: number,
  position: number,
  request: UpdateReviewDiffRequest,
): Promise<ReviewResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}`,
    request,
  );

  return await handleResponse(response, ReviewResource);
}

export async function updateReview(
  owner: string,
  repo: string,
  number: number,
  request: UpdateReviewRequest,
): Promise<ReviewResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}`,
    request,
  );

  return await handleResponse(response, ReviewResource);
}

export async function publishReview(
  owner: string,
  repo: string,
  number: number,
  request: PublishReviewRequest,
): Promise<ReviewResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/publish`,
    request,
  );

  return await handleResponse(response, ReviewResource);
}

export async function publishReviewDiff(
  owner: string,
  repo: string,
  number: number,
  position: number,
): Promise<ReviewResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}/publish`,
    {},
  );

  return await handleResponse(response, ReviewResource);
}

export async function reviewReviewDiff(
  owner: string,
  repo: string,
  number: number,
  position: number,
  request: ReviewReviewDiffRequest,
): Promise<ReviewResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}/review`,
    request,
  );

  return await handleResponse(response, ReviewResource);
}

export async function mergeDiff(
  owner: string,
  repo: string,
  number: number,
  position: number,
): Promise<ReviewResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}/merge`,
    {},
  );

  return await handleResponse(response, ReviewResource);
}

export async function updateReviewComment(
  owner: string,
  repo: string,
  number: number,
  commentId: string,
  request: UpdateReviewCommentRequest,
): Promise<ReviewCommentResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/comment/${commentId}`,
    request,
  );

  return await handleResponse(response, ReviewCommentResource);
}

export async function resolveReviewComment(
  owner: string,
  repo: string,
  number: number,
  commentId: string,
  resolved: boolean,
): Promise<ReviewCommentResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/comment/${commentId}/resolve`,
    { resolved },
  );

  return await handleResponse(response, ReviewCommentResource);
}

export async function replyToReviewComment(
  owner: string,
  repo: string,
  number: number,
  commentId: string,
  request: ReplyToReviewCommentRequest,
): Promise<ReviewCommentResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/comment/${commentId}/reply`,
    request,
  );

  return await handleResponse(response, ReviewCommentResource);
}

File: gitdot-server/src/handler/review.rs
mod add_review_reviewer;
mod get_review;
mod get_review_diff_blobs;
mod list_reviews;
mod merge_review_diff;
mod publish_review;
mod publish_review_diff;
mod remove_review_reviewer;
mod reply_to_review_comment;
mod resolve_review_comment;
mod review_review_diff;
mod update_review;
mod update_review_comment;
mod update_review_diff;

use crate::app::AppState;
use axum::{
    Router,
    routing::{delete, get, patch, post},
};

use add_review_reviewer::add_review_reviewer;
use get_review::get_review;
use get_review_diff_blobs::get_review_diff_blobs;
use list_reviews::list_reviews;
use merge_review_diff::merge_review_diff;
use publish_review::publish_review;
use publish_review_diff::publish_review_diff;
use remove_review_reviewer::remove_review_reviewer;
use reply_to_review_comment::reply_to_review_comment;
use resolve_review_comment::resolve_review_comment;
use review_review_diff::review_review_diff;
use update_review::update_review;
use update_review_comment::update_review_comment;
use update_review_diff::update_review_diff;

pub fn create_review_router() -> Router<AppState> {
    Router::new()
        .route(
            "/repository/{owner}/{repo}/review/{number}",
            get(get_review).patch(update_review),
        )
        .route("/repository/{owner}/{repo}/reviews", get(list_reviews))
        .route(
            "/repository/{owner}/{repo}/review/{number}/publish",
            post(publish_review),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/reviewer",
            post(add_review_reviewer),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/reviewer/{reviewer_name}",
            delete(remove_review_reviewer),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}",
            patch(update_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/blobs",
            get(get_review_diff_blobs),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/merge",
            post(merge_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/publish",
            post(publish_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/review",
            post(review_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/comment/{comment_id}",
            patch(update_review_comment),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/comment/{comment_id}/reply",
            post(reply_to_review_comment),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/comment/{comment_id}/resolve",
            post(resolve_review_comment),
        )
}
